actors/miner/tests/miner_actor_test_wpost.rs

```diff
@@ -663,28 +671,570 @@ fn duplicate_proof_rejected_with_many_partitions() {
+    let dlinfo = h.advance_to_deadline(&mut rt, dlidx);
+
+    // Submit PoSt
+    let partition = miner::PoStPartition { index: pidx, skipped: make_empty_bitfield() };
+    h.submit_window_post(
+        &mut rt,
+        &dlinfo,
+        vec![partition],
+        infos,
+        PoStConfig::with_expected_power_delta(&pwr),
+    );
+
+    // Dispute it.
+    let params = miner::DisputeWindowedPoStParams { deadline: dlinfo.index, post_index: 0 };
+
+    rt.set_caller(*ACCOUNT_ACTOR_CODE_ID, h.worker);
+    rt.expect_validate_caller_type(vec![*ACCOUNT_ACTOR_CODE_ID, *MULTISIG_ACTOR_CODE_ID]);
+    h.expect_query_network_info(&mut rt);
+
+    let result = rt.call::<miner::Actor>(
+        miner::Method::DisputeWindowedPoSt as u64,
+        &RawBytes::serialize(params).unwrap(),
+    );
+    expect_abort_contains_message(
+        ExitCode::ErrForbidden,
+        "can only dispute window posts during the dispute window",
+        result,
+    );
+    rt.verify();
+}
 
 #[test]
-fn can_dispute_up_till_window_end_but_not_after() {}
+fn can_dispute_up_till_window_end_but_not_after() {
+    let period_offset = ChainEpoch::from(100);
+    let precommit_epoch = ChainEpoch::from(1);
+
+    let mut h = ActorHarness::new(period_offset);
+    h.set_proof_type(RegisteredSealProof::StackedDRG2KiBV1P1);
+
+    let mut rt = h.new_runtime();
+    rt.epoch = precommit_epoch;
+    rt.balance.replace(TokenAmount::from(BIG_BALANCE));
+
+    h.construct_and_verify(&mut rt);
+
+    let infos = h.commit_and_prove_sectors(&mut rt, 1, DEFAULT_SECTOR_EXPIRATION, vec![], true);
+    let sector = infos[0].clone();
+
+    let state = h.get_state(&rt);
+    let (dlidx, _) = state.find_sector(&rt.policy, &rt.store, sector.sector_number).unwrap();
+
+    let nextdl = miner::DeadlineInfo::new(
+        state.proving_period_start,
+        dlidx,
+        rt.epoch,
+        rt.policy.wpost_period_deadlines,
+        rt.policy.wpost_proving_period,
+        rt.policy.wpost_challenge_window,
+        rt.policy.wpost_challenge_lookback,
+        rt.policy.fault_declaration_cutoff,
+    )
+    .next_not_elapsed();
+
+    h.advance_and_submit_posts(&mut rt, &infos);
+    let window_end = nextdl.close + rt.policy.wpost_dispute_window;
+
+    // first, try to dispute right before the window end.
+    // We expect this to fail "normally" (fail to disprove).
+    rt.epoch = window_end - 1;
+    h.dispute_window_post(&mut rt, &nextdl, 0, &infos, None);
+
+    // Now set the epoch at the window end. We expect a different error.
+    rt.epoch = window_end;
+
+    // Now try to dispute.
+    let params = miner::DisputeWindowedPoStParams { deadline: dlidx, post_index: 0 };
+    rt.set_caller(*ACCOUNT_ACTOR_CODE_ID, h.worker);
+    rt.expect_validate_caller_type(vec![*ACCOUNT_ACTOR_CODE_ID, *MULTISIG_ACTOR_CODE_ID]);
+
+    let current_reward = reward::ThisEpochRewardReturn {
+        this_epoch_baseline_power: h.baseline_power,
```

Comment:
nit: I think you have a helper function on the harness for expecting these two network query messages